op-challenger/fault/service.go

```diff
@@ -82,16 +88,30 @@ func newTypedService(ctx context.Context, logger log.Logger, cfg *config.Config,
 		return nil, fmt.Errorf("failed to bind the fault contract: %w", err)
 	}
 
-	agent := NewAgent(loader, cfg.GameDepth, provider, responder, updater, cfg.AgreeWithProposedOutput, gameLogger)
-
 	return &service{
-		agent:                   agent,
+		agent:                   NewAgent(loader, cfg.GameDepth, provider, responder, updater, cfg.AgreeWithProposedOutput, gameLogger),
 		agreeWithProposedOutput: cfg.AgreeWithProposedOutput,
 		caller:                  caller,
 		logger:                  gameLogger,
 	}, nil
 }
 
+// ValidateAbsolutePrestate validates the absolute prestate of the fault game.
+func ValidateAbsolutePrestate(ctx context.Context, trace types.TraceProvider, loader Loader) error {
+	providerPrestate, err := trace.AbsolutePreState(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get the trace provider's absolute prestate: %w", err)
+	}
+	onchainPrestate, err := loader.FetchAbsolutePrestate(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get the onchain absolute prestate: %w", err)
+	}
+	if !bytes.Equal(providerPrestate, onchainPrestate) {
+		return fmt.Errorf("trace provider's absolute prestate does not match onchain absolute prestate")
+	}
```

Comment:
This will always fail beacause you're comparing the absolute pre state pre-image (from the trace provider) to the absolute pre state hash (from the contract).  Which I think is why all the e2e tests are failing. :)